### DISCUSSION TITLE
the position of words render by textlayer are not the same with canvas

### DISCUSSION THREAD

**Discussion by zishengwu** [mozilla/pdf.js#18068] [Q&A]
Attach (recommended) or Link to PDF file here: [](https://mozilla.github.io/pdf.js/web/compressed.tracemonkey-pldi-09.pdf)

Configuration:
- Web browser and its version:  Chrome 124.0.6367.119(arm 64)
- Operating system and its version: macOS 14.1.1 (23B81)
- PDF.js version: 4.2.67  (from npm: pdfjs-dist)
- Is a browser extension: No

Steps to reproduce the problem:
1. render pdf success, and the canvas can be found.
2. add textlayer to render, but the position of those words is not correct.

What is the expected behavior? (add screenshot)
![image](https://github.com/mozilla/pdf.js/assets/49832481/b1a6e934-2fa2-4d8f-9f0b-909486868af2)
when i zoomin the browser up to 250%, seems the position correct.

What went wrong? (add screenshot)
![image](https://github.com/mozilla/pdf.js/assets/49832481/f3c9b6f6-1252-457c-bd02-dc39a323dc78)

Link to a viewer (if hosted on a site other than mozilla.github.io/pdf.js or as Firefox/Chrome extension):

Here is my code.
![image](https://github.com/mozilla/pdf.js/assets/49832481/6b019869-46ae-49e9-93ec-2beebc2cee67)

Here is the detail from browser 
![image](https://github.com/mozilla/pdf.js/assets/49832481/35f8f4ed-de28-45ae-b733-ca445d352e33)

can anyone help me?

**Comment by cpernsteineribm** [comment]
see https://github.com/mozilla/pdf.js/issues/17561

**Comment by lfoppiano** [comment]
Hi @zishengwu, thanks to your example which seems to provide an update version of programatically using pdf.js I could get to add the text layers. I have the same issue as you and I believe it might be due to the scaling. 

For example the url at the top is appearing, as text in the highlighted portion in this example: 

<img width="741" alt="image" src="https://github.com/mozilla/pdf.js/assets/15426/6fa125b0-04d4-4e2b-b60f-8e073873bd82">

I'm totally lost on how to fix it, though. 

I've opened #18112 just yesteday before stumbling upon this discussion, today 👍

**Reply by zishengwu** [reply]
`import 'pdfjs-dist/web/pdf_viewer.css';

async function renderFullPage(){
        const pdf = await pdfjsLib.getDocument(src).promise;

        const pdfContainer = document.createElement('div');
        pdfContainer.style.setProperty('--scale-factor', scale_ratio);

        for (let i=1; i<=pdf.numPages; i++){
            const pageNumber = i;
            const page = await pdf.getPage(pageNumber);

            const viewport = page.getViewport({scale: scale_ratio});

            const canvas = document.createElement('canvas');
            const context = canvas.getContext('2d');

            canvas.height = viewport.height ;
            canvas.width = viewport.width ;
            
            const renderContext = {
                canvasContext: context,
                viewport: viewport
            };
        
            await page.render(renderContext);

      
            const canvasWrapper = document.createElement('div');
            canvasWrapper.className = 'canvasWrapper';
            canvasWrapper.appendChild(canvas);

            const textContent = await page.getTextContent();
            const textLayerDiv = document.createElement('div');
  
            textLayerDiv.className = `textLayer`;

            pdfjsLib.renderTextLayer({
                textContentSource: textContent,
                container: textLayerDiv,
                viewport: viewport,
                textDivs: []
            });
            
            const pageDiv = document.createElement('div');
            pageDiv.className = 'page';
          
            pageDiv.style = "position: relative; margin-bottom:10px";

            pageDiv.appendChild(canvasWrapper);
            pageDiv.appendChild(textLayerDiv);
            
            pdfContainer.appendChild(pageDiv);
            
        }

    }`
    
    Hi, maybe you can try this code. After set the `--scale-factor` in top div, it works for me now.

**Reply by lfoppiano** [reply]
Thanks.  I think I integrated correctly, but now I have an issue with pdf.js version 3 and I'm struggling to update to version 4.

**Reply by lfoppiano** [reply]
@zishengwu Thanks! how do you calculate the scale ratio?

**Reply by zishengwu** [reply]
I don't have a specific scale ratio, but according to some codes from Internet, `scale_ratio=1.5` is a experience value

**Reply by lfoppiano** [reply]
I finally solved, I calculate the scale dividing the real width of the PDF divided by the unscaled viewport. The code is here: https://github.com/lfoppiano/streamlit-pdf-viewer/pull/52 

BTW I noticed that when I enable the text layers they might interfere with the annotations that I'm adding on top of the documents.
